How to add more node to the existing deployed cluster for OnPrem Deployment (#7406)
chef/automate: components/docs-chef-io/content/automate/ha_onprim_deployment_procedure.md

-   `opensearch_private_ips` Eg: ["192.0.2.1", "192.0.2.2", "192.0.2.2"]
-   `postgresql_private_ips` Eg: ["192.0.3.1", "192.0.3.2", "192.0.3.2"]

### How To Add More Nodes to the Onprem Deployment 
- Open the `config.toml` at bastion node.

> - change the `instance_count` value, as explain in below example.

Reviewer: ```suggestion
- Change the `instance_count` value, as explained in below example.
```

- Add the `Ip address` for the respetive node at the end, as expain in the example.
- Make sure that all the necesssary port and fire wall setting are allign in the new node.

For example : Add new Automate node to the existing deployed cluster.
  | Old Config | => | New Config |
  | :--- | :--- | :--- |
  | [automate.config] <br> instance_count = "1" <br> [existing_infra.config] <br> automate_private_ips = ["10.0.1.0"] |  | [automate.config] <br> instance_count = "2" <br> [existing_infra.config] <br> automate_private_ips = ["10.0.1.0","10.0.2.0"] |

- Trigger the deployment command again from the bastion node.
- In below deploy command,latest.aib will replace with your airgap bundle name which is running on the current cluster.
  ```
    chef-automate deploy config.toml --airgap-bundle latest.aib
  ```
- Above process can be done for `chef-server`, `postgresql` and `opensearch` cluster as well
- In case of Deployment failed please refer the troubleshoot document [here](/automate/ha_onprim_deployment_procedure/#Troubleshooting).

### How To Remove Any Nodes From Frontend Cluster (Onprem Deployment). 
{{< warning >}}

>  We can not remove the node from backend cluster, but we can replace the node. For replace a node please refer [this](/automate/ha_onprim_deployment_procedure/#How-to-Replace-Node-in-Automate-HA-Cluster).

Reviewer: ```suggestion
 We do not recommend removal of any node from backend cluster, but replacement of node is recommended. For replacement of a node please refer: [Replace Node in Automate HA](/automate/ha_onprim_deployment_procedure/#How-to-Replace-Node-in-Automate-HA-Cluster).
```

{{< /warning >}}

- Open the `config.toml` at bastion node.

> - change the `instance_count` value, as explain in below example.

Reviewer: ```suggestion
- Change the `instance_count` value, as explained in below example.
```

- Remove the `Ip address` for the respetive node, as explain in the example.

For example : Remove Automate node to the existing deployed cluster.
  | Old Config | => | New Config |
  | :--- | :--- | :--- |
  | [automate.config] <br> instance_count = "3" <br> [existing_infra.config] <br> automate_private_ips = ["10.0.1.0","10.0.2.0","10.0.3.0"] |  |[automate.config] <br> instance_count = "2" <br> [existing_infra.config] <br> automate_private_ips = ["10.0.1.0","10.0.3.0"] |
  
- Trigger the deployment command again from the bastion node.

> - In below deploy command,`latest.aib` will replace with your airgap bundle name which is running on the current cluster.

Reviewer: ```suggestion
- In below deploy command, `latest.aib` will replace with your airgap bundle name which is running on the current cluster.
```

- In case of Deployment failed please refer the troubleshoot document [here](/automate/ha_onprim_deployment_procedure/#Troubleshooting).

### How to Replace Node in Automate HA Cluster
- Open the `config.toml` at bastion node.
- Remove the `Ip Address` of unhealth node as explain in below example.
- Add the new `Ip address` value, as explain in below example.
- Make sure that all the necesssary port and fire wall setting are allign in the new node.

 For example : Remove node to the existing deployed Postgres cluster.
 In the example we are replacing the node `10.0.7.0` with `10.0.9.0`.

  | Old Config | => | New Config |
  | :--- | :--- | :--- |
  | [existing_infra.config] <br> postgresql_private_ips = ["10.0.6.0","10.0.7.0","10.0.8.0"] |  | [existing_infra.config] <br> postgresql_private_ips = ["10.0.6.0","10.0.9.0","10.0.8.0"] |
  
- Run the below command from the bastion node.
  ```
  cd /hab/a2_deploy_workspace/terraform
  for x in $(terraform state list -state=/hab/a2_deploy_workspace/terraform/terraform.tfstate | grep module); do terraform taint $x; done
  cd -
  ``` 
- Run the `deploy` command to Add new node into the cluster.

> - In below deploy command,latest.aib will replace with your airgap bundle name which is running on the current cluster.

Reviewer: ```suggestion
- In below deploy command, `latest.aib` will replace with your airgap bundle name which is running on the current cluster.
```

  ```
    chef-automate deploy config.toml --airgap-bundle latest.aib
  ```

### Troubleshooting

> - while adding new node into the cluster if we get an below error

Reviewer: ```suggestion
#### Failure in adding nodes
While adding new nodes into the cluster if we get an below error
```

  ```
  Error: Upload failed: scp: /var/automate-ha: Permission denied
  ```

> - Resolution : Execute the below command.

Reviewer: ```suggestion
**Resolution** : Execute the below command.
```

  ```
  cd /hab/a2_deploy_workspace/terraform
  for x in $(terraform state list -state=/hab/a2_deploy_workspace/terraform/terraform.tfstate | grep module); do terraform taint $x; done
  cd -
   ``` 

> - Once the module's tainted, run the `deploy` command again `chef-automate deploy config.toml --airgap-bundle latest.aib`

Reviewer: ```suggestion
- Once the module's tainted, run the deploy command again:
    ```bash
    chef-automate deploy config.toml --airgap-bundle latest.aib
    ```
```

>  (`latest.aib` will replace with your airgap bundle name which is running on the current cluster).

Reviewer: ```suggestion
 (`latest.aib` should be replaced with your airgap bundle name which is running on the current cluster).
```
This should be moved before we show the deploy command.
